core/drivers/hisi_trng.c

```diff
@@ -0,0 +1,84 @@
+// SPDX-License-Identifier: GPL-2.0
+/* Copyright (c) 2021 HiSilicon Limited. */
+
+#include <initcall.h>
+#include <io.h>
+#include <malloc.h>
+#include <mm/core_memprot.h>
+#include <platform_config.h>
+#include <rng_support.h>
+
+#define HTRNG_RANDATA_REG 0xF0
+#define HTRNG_BYTES 4
+
+#define POLL_PERIOD 10
+#define POLL_TIMEOUT 1000
+
+struct hisi_trng {
+	vaddr_t base;
+	uint32_t ver;
+};
+
+union trng_fifo {
+	uint32_t val;
+	uint8_t data_byte[HTRNG_BYTES];
+};
+
+static struct hisi_trng *trng_dev;
+
+TEE_Result hw_get_random_bytes(void *buf, size_t len)
+{
+	TEE_Result ret = TEE_ERROR_GENERIC;
+	uint8_t *out = (uint8_t *)buf;
+	static union trng_fifo fifo = { };
```

Comment:
You use dynamic memory allocation to keep track of the base address which is more or less static, but static memory allocation to keep the dynamic part of the state.

I'd prefer a more direct approach with global:
``` C
static struct hisi_trng trng_dev; /* or perhaps just "static vaddr_t trng_base" */
static union trng_fifo trng_fifo;
static unsigned int trng_lock = SPINLOCK_UNLOCK
```